### PR TITLE
NEWTAG-587 input() function

### PR DESCRIPTION
**Pull request by thpr**
This is the input() function for the new formula system, including the STATSCORE code control (which enables the use of the STATSCORE channels)

### CODE REVIEW

File: code/pluginbuild.xml [hunk 1]
```diff
@@ -6229,6 +6229,13 @@ jar-all-plugins - Generate the plugin jar files
 				</patternset>
 			</fileset>
 		</jar>
+		<jar jarfile="${systemlstplugins.dir}/GameMode-CodeControlToken-STATSCORE.jar" manifest="${src.java.dir}/plugin/lsttokens/manifest.mf">
```
Comment (grimreaper) [inline code/pluginbuild.xml]: is this auto-generated? If not, can we make it so? Its hard to review and seems to be largely redundant.

File: code/src/java/pcgen/output/channel/ChannelCompatibility.java [hunk 2]
```diff
@@ -34,7 +34,8 @@
 		if (ControlUtilities.hasControlToken(Globals.getContext(),
 			CControl.STATSCORE))
 		{
-			return (WriteableReferenceFacade<Integer>) ChannelUtilities.getChannel(id, stat, CControl.STATSCORE);
+			return (WriteableReferenceFacade<Integer>) ChannelUtilities
```
Comment (grimreaper) [inline code/src/java/pcgen/output/channel/ChannelCompatibility.java]: is the cast needed?

File: code/src/java/pcgen/output/channel/ChannelUtilities.java [hunk 3]
```diff
@@ -35,14 +37,44 @@
...
 {
-	private static final VariableLibraryFacet VARLIB_FACET = FacetLibrary
-		.getFacet(VariableLibraryFacet.class);
-	private static final ScopeFacet SCOPE_FACET = FacetLibrary
-		.getFacet(ScopeFacet.class);
-	private static final VariableStoreFacet RESULT_FACET = FacetLibrary
-		.getFacet(VariableStoreFacet.class);
-	private static final SolverManagerFacet MGR_FACET = FacetLibrary
-		.getFacet(SolverManagerFacet.class);
+	private static final VariableLibraryFacet VARLIB_FACET =
```
Comment (grimreaper) [inline code/src/java/pcgen/output/channel/ChannelUtilities.java]: if I wasn't guilty of the same things I'd get annoyed at formatting and code changes in the same commit :-p

File: code/src/java/pcgen/output/channel/ChannelUtilities.java [hunk 4]
```diff
@@ -35,14 +37,44 @@
...
+	 * @param owner
+	 *            The owning object of the Channel
+	 * @param name
+	 *            The name of the channel
+	 * @param formatManager
+	 *            The FormatManager for the channel
+	 * @return A Channel for the given CharID, owning object, and name of the
+	 *         channel
+	 */
+	public static <T> VariableChannel<T> generateChannel(CharID id,
```
Comment (grimreaper) [inline code/src/java/pcgen/output/channel/ChannelUtilities.java]: no restrictions on T ?

File: code/src/java/plugin/function/InputFunction.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	public final FormatManager<?> allowArgs(SemanticsVisitor visitor, Node[] args,
+		FormulaSemantics semantics)
+	{
+		int argCount = args.length;
+		if (argCount != 1)
+		{
+			semantics.setInvalid("Function " + getFunctionName()
+				+ " received incorrect # of arguments, expected: 1 got "
+				+ args.length + " " + Arrays.asList(args));
+			return null;
```
Comment (grimreaper) [inline code/src/java/plugin/function/InputFunction.java]: prefer Optional instead of null.

File: code/src/java/pcgen/output/channel/ChannelCompatibility.java (hunk 2 above)
Comment (cpmeister) [inline code/src/java/pcgen/output/channel/ChannelCompatibility.java]: If you can pull it off without the cast go for it

File: code/src/java/plugin/function/InputFunction.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public final FormatManager<?> allowArgs(SemanticsVisitor visitor, Node[] args,
+		FormulaSemantics semantics)
+	{
+		int argCount = args.length;
```
Comment (grimreaper) [inline code/src/java/plugin/function/InputFunction.java]: Can we make some kind of checkArguments function and call down instead of having separate checks inside of a monolithic check

File: code/src/java/plugin/function/InputFunction.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		if (argCount != 1)
+		{
+			semantics.setInvalid("Function " + getFunctionName()
+				+ " received incorrect # of arguments, expected: 1 got "
+				+ args.length + " " + Arrays.asList(args));
+			return null;
+		}
+		//String node (name)
+		Node inputNode = args[0];
+		if (!(inputNode instanceof ASTQuotString))
```
Comment (grimreaper) [inline code/src/java/plugin/function/InputFunction.java]: :( in new code

File: code/src/java/plugin/function/InputFunction.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			return null;
+		}
+		ASTQuotString quotString = (ASTQuotString) inputNode;
+		String inputName = quotString.getText();
+		String varName = ChannelUtilities.createVarName(inputName);
+		VariableLibrary varLib =
+				semantics.peek(FormulaSemantics.FMANAGER).getFactory();
+		LegalScope scope = semantics.peek(FormulaSemantics.SCOPE);
+		FormatManager<?> formatManager =
+				varLib.getVariableFormat(scope, varName);
```
Comment (grimreaper) [inline code/src/java/plugin/function/InputFunction.java]: can getFormaulFormat become an Optional instead?

File: code/src/utest/plugin/function/InputFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+		String formula = "input(\"STR\")";
+		SimpleNode node = TestUtilities.doParse(formula);
+		isValid(formula, node, numberManager, null);
+		isStatic(formula, node, false);
+		evaluatesTo(formula, node, Integer.valueOf(0));
+		strChannel.set(2);
+		evaluatesTo(formula, node, Integer.valueOf(2));
+		Object rv =
+				new ReconstructionVisitor().visit(node, new StringBuilder());
+		assertTrue(rv.toString().equals(formula));
```
Comment (grimreaper) [inline code/src/utest/plugin/function/InputFunctionTest.java]: assertEquals

File: code/pluginbuild.xml (hunk 1 above)
Comment (thpr) [inline code/pluginbuild.xml]: Not auto generated, but happy to go there if you want to do that as a separate PR

File: code/src/java/pcgen/output/channel/ChannelUtilities.java (hunk 4 above)
Comment (thpr) [inline code/src/java/pcgen/output/channel/ChannelUtilities.java]: No.  Channel could be a String, Number, anything

File: code/src/java/plugin/function/InputFunction.java (hunk 5 above)
Comment (thpr) [inline code/src/java/plugin/function/InputFunction.java]: So the entire concept of how allowArgs works depends on returning null... if you want to change that, it's a -formula library change we can discuss

File: code/src/java/plugin/function/InputFunction.java (hunk 8 above)
Comment (thpr) [inline code/src/java/plugin/function/InputFunction.java]: basically, not in this checkin, the system is designed a certain way that relies on nulls

File: code/src/java/plugin/function/InputFunction.java (hunk 6 above)
Comment (thpr) [inline code/src/java/plugin/function/InputFunction.java]: Explain further what you mean, pls
